elixir/lib/extism/plugin.ex

```diff
@@ -0,0 +1,58 @@
+defmodule Extism.Plugin do
+  defstruct [
+    # The actual NIF Resource.
+    resource: nil,
+    # Normally the compiler will happily do stuff like inlining the
+    # resource in attributes. This will convert the resource into an
+    # empty binary with no warning. This will make that harder to
+    # accidentaly do.
+    # It also serves as a handy way to tell plugins apart.
+    reference: nil
+  ]
+
+  def wrap_resource(resource) do
+    %__MODULE__{
+      resource: resource,
+      reference: make_ref()
+    }
+  end
+
+  def new(manifest, wasi \\ false) when is_map(manifest) do
+    {:ok, manifest_payload} = JSON.encode(manifest)
```

Comment:
Instead of sharing the Manifest struct, which rustler does allow with some magic, I decided to just encode a map to json and send it as a string. I'm open to other ways of doing this. One downside with this is you need the validation in rust land and you would probably catch less problems at elixir compile time. The upside is less interface coupling to this language.

Reply:
Ok, thanks for calling this out! Let me think about it and test around a bit

Reply:
I think this is probably a fine solution for the time being - especially if we think the `Manifest` type itself will be in flux until a more stable version of Extism in the future. 

It's probably worth exploring how we'd define this type in `witx` format, and try to use (and contribute to) a tool such as: 

https://github.com/bytecodealliance/wit-bindgen or https://github.com/jedisct1/witx-codegen